{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene data cooking\n",
    "\n",
    "* gene_table.txt contains summary annotation on all human genes, based on the Ensembl annotation:\n",
    "* gene_name\n",
    "* gene_biotype (e.g., protein_coding, pseudogene, lincRNA, miRNA etc)\n",
    "* chromosome (on which the gene is located\n",
    "* strand (on which the gene is located)\n",
    "* transcript_count (the number of known isoforms of the gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reading from CSV files\n",
    "\n",
    "* Pandas provides the read_csv() function which can be used data from a CSV file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# by read from file gene_table and delimit the fields using ','\n",
    "df=pd.read_csv('gene_table.txt', sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Accessing the DataFrame\n",
    "\n",
    "## Displaying the DataFame's top rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr20</td>\n",
       "      <td>-</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name    gene_biotype chromosome strand  transcript_count\n",
       "0    TSPAN6  protein_coding       chrX      -                 5\n",
       "1      TNMD  protein_coding       chrX      +                 2\n",
       "2      DPM1  protein_coding      chr20      -                 6\n",
       "3     SCYL3  protein_coding       chr1      -                 5\n",
       "4  C1orf112  protein_coding       chr1      +                 9"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# head() method displays by default the top-5 rows of a dataframe\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  gene_name    gene_biotype chromosome strand  transcript_count\n",
      "0    TSPAN6  protein_coding       chrX      -                 5\n",
      "1      TNMD  protein_coding       chrX      +                 2\n",
      "2      DPM1  protein_coding      chr20      -                 6\n",
      "3     SCYL3  protein_coding       chr1      -                 5\n",
      "4  C1orf112  protein_coding       chr1      +                 9\n",
      "5       FGR  protein_coding       chr1      -                 7\n",
      "6       CFH  protein_coding       chr1      +                 6\n",
      "7     FUCA2  protein_coding       chr6      -                 3\n",
      "8      GCLC  protein_coding       chr6      -                13\n",
      "9      NFYA  protein_coding       chr6      +                 2\n"
     ]
    }
   ],
   "source": [
    "# head() accepts also a number which configures the number of top rows of the dataframe to be displayed\n",
    "print(df.head(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Displaying the DataFrame's last rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51322</th>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>chrMT</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51323</th>\n",
       "      <td>CRYM-AS1</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51324</th>\n",
       "      <td>RP11-100E13.1</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51325</th>\n",
       "      <td>VTRNA2-1</td>\n",
       "      <td>vaultRNA</td>\n",
       "      <td>chr5</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51326</th>\n",
       "      <td>BCYRN1</td>\n",
       "      <td>scRNA</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name                   gene_biotype chromosome strand  \\\n",
       "51322        MT-RNR1                        Mt_rRNA      chrMT      +   \n",
       "51323       CRYM-AS1  bidirectional_promoter_lncRNA      chr16      +   \n",
       "51324  RP11-100E13.1  bidirectional_promoter_lncRNA       chr1      -   \n",
       "51325       VTRNA2-1                       vaultRNA       chr5      -   \n",
       "51326         BCYRN1                          scRNA       chr2      +   \n",
       "\n",
       "       transcript_count  \n",
       "51322                 1  \n",
       "51323                 3  \n",
       "51324                 1  \n",
       "51325                 1  \n",
       "51326                 1  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# similarly tail() displays the last 5 rows of a dataframe\n",
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51324</th>\n",
       "      <td>RP11-100E13.1</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51325</th>\n",
       "      <td>VTRNA2-1</td>\n",
       "      <td>vaultRNA</td>\n",
       "      <td>chr5</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51326</th>\n",
       "      <td>BCYRN1</td>\n",
       "      <td>scRNA</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name                   gene_biotype chromosome strand  \\\n",
       "51324  RP11-100E13.1  bidirectional_promoter_lncRNA       chr1      -   \n",
       "51325       VTRNA2-1                       vaultRNA       chr5      -   \n",
       "51326         BCYRN1                          scRNA       chr2      +   \n",
       "\n",
       "       transcript_count  \n",
       "51324                 1  \n",
       "51325                 1  \n",
       "51326                 1  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# or a different amount if an argument is specified\n",
    "df.tail(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieving the DataFrame's dimensions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame dimensions:  (51327, 5)\n",
      "DataFrame number of rows:  51327\n",
      "DataFrame number of columns:  5\n"
     ]
    }
   ],
   "source": [
    "print(\"DataFrame dimensions: \", df.shape)\n",
    "\n",
    "# the number of rows\n",
    "print(\"DataFrame number of rows: \", df.shape[0])\n",
    "\n",
    "# the number of columns\n",
    "print(\"DataFrame number of columns: \", df.shape[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selecting DataFrame partitions\n",
    "\n",
    "### Selecting DataFrame rows\n",
    "\n",
    "- single rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr20</td>\n",
       "      <td>-</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name    gene_biotype chromosome strand  transcript_count\n",
       "0    TSPAN6  protein_coding       chrX      -                 5\n",
       "1      TNMD  protein_coding       chrX      +                 2\n",
       "2      DPM1  protein_coding      chr20      -                 6\n",
       "3     SCYL3  protein_coding       chr1      -                 5\n",
       "4  C1orf112  protein_coding       chr1      +                 9"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* locate rows: loc[] when using row label/index, iloc[] when using index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row with label 11 of DataFrame:\n",
      "gene_name                   NIPAL3\n",
      "gene_biotype        protein_coding\n",
      "chromosome                    chr1\n",
      "strand                           +\n",
      "transcript_count                 7\n",
      "Name: 11, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# retrieving DataFrame row by row index\n",
    "print(\"Row with label 11 of DataFrame:\")\n",
    "print(df.loc[11])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_name                    LAS1L\n",
      "gene_biotype        protein_coding\n",
      "chromosome                    chrX\n",
      "strand                           -\n",
      "transcript_count                 5\n",
      "Name: 12, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# retrieving DataFrame row by index\n",
    "#print(\"13th row:\")\n",
    "print(df.iloc[12])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- multiple rows by slicing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Displaying rows between label 8 and 11\n",
      "\n",
      "   gene_name    gene_biotype chromosome strand  transcript_count\n",
      "8       GCLC  protein_coding       chr6      -                13\n",
      "9       NFYA  protein_coding       chr6      +                 2\n",
      "10     STPG1  protein_coding       chr1      -                12\n",
      "11    NIPAL3  protein_coding       chr1      +                 7\n"
     ]
    }
   ],
   "source": [
    "# using a slice in 'loc', returns the part of the DataFrame, between the rows given in the slice\n",
    "print(\"Displaying rows between label 8 and 11\")\n",
    "print()\n",
    "print(df.loc[8:11])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Displaying rows whose index>=8  and index<11:\n",
      "\n",
      "   gene_name    gene_biotype chromosome strand  transcript_count\n",
      "8       GCLC  protein_coding       chr6      -                13\n",
      "9       NFYA  protein_coding       chr6      +                 2\n",
      "10     STPG1  protein_coding       chr1      -                12\n"
     ]
    }
   ],
   "source": [
    "# using a slice in iloc, returns the part of the DataFrame, indexed by the numbers in the slice\n",
    "# (same idea as in Python lists)\n",
    "print(\"Displaying rows whose index>=8  and index<11:\")\n",
    "print()\n",
    "print(df.iloc[8:11])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selecting DataFrame columns\n",
    "\n",
    "- single column (returns a Series)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0               TSPAN6\n",
      "1                 TNMD\n",
      "2                 DPM1\n",
      "3                SCYL3\n",
      "4             C1orf112\n",
      "             ...      \n",
      "51322          MT-RNR1\n",
      "51323         CRYM-AS1\n",
      "51324    RP11-100E13.1\n",
      "51325         VTRNA2-1\n",
      "51326           BCYRN1\n",
      "Name: gene_name, Length: 51327, dtype: object\n"
     ]
    }
   ],
   "source": [
    "# retrieving gene_name as a key value\n",
    "print(df['gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No column was found\n"
     ]
    }
   ],
   "source": [
    "# get method can also be used to make sure no error is raised when the given column does not exist\n",
    "print(df.get('gene-name', 'No column was found'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- selecting multiple columns (returns a new DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           gene_name strand  transcript_count\n",
      "0             TSPAN6      -                 5\n",
      "1               TNMD      +                 2\n",
      "2               DPM1      -                 6\n",
      "3              SCYL3      -                 5\n",
      "4           C1orf112      +                 9\n",
      "...              ...    ...               ...\n",
      "51322        MT-RNR1      +                 1\n",
      "51323       CRYM-AS1      +                 3\n",
      "51324  RP11-100E13.1      -                 1\n",
      "51325       VTRNA2-1      -                 1\n",
      "51326         BCYRN1      +                 1\n",
      "\n",
      "[51327 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "# retrieving multiple columns of a DataFrame\n",
    "print(df[['gene_name', 'strand','transcript_count']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Iterating through a DataFrame's rows\n",
    "* The iterrows() method generates an iterator object of the DataFrame, allowing us to iterate each row in the DataFrame.\n",
    "* Each iteration produces an index object and a row object (a Pandas Series object)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row label:  0\n",
      "Row data:  gene_name                   TSPAN6\n",
      "gene_biotype        protein_coding\n",
      "chromosome                    chrX\n",
      "strand                           -\n",
      "transcript_count                 5\n",
      "Name: 0, dtype: object\n",
      "---------\n",
      "Row label:  1\n",
      "Row data:  gene_name                     TNMD\n",
      "gene_biotype        protein_coding\n",
      "chromosome                    chrX\n",
      "strand                           +\n",
      "transcript_count                 2\n",
      "Name: 1, dtype: object\n",
      "---------\n",
      "Row label:  2\n",
      "Row data:  gene_name                     DPM1\n",
      "gene_biotype        protein_coding\n",
      "chromosome                   chr20\n",
      "strand                           -\n",
      "transcript_count                 6\n",
      "Name: 2, dtype: object\n",
      "---------\n"
     ]
    }
   ],
   "source": [
    "#get first 3 rows of the dataframe\n",
    "tiny_df = df.iloc[0:3] # 0, 1, 2\n",
    "#iterate over each row\n",
    "# iterrows returns an index and a panda series\n",
    "for row_label, row_data in tiny_df.iterrows():\n",
    "    # row_label: label of each row\n",
    "    # row_data: data of each row (a Series)\n",
    "    print(\"Row label: \", row_label)\n",
    "    print(\"Row data: \", row_data)\n",
    "    print(\"---------\")\n",
    "    #print(\"Row data: \", row_data['chromosome'])\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieving statistics\n",
    "\n",
    "- by using statistical methods provided by pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0        5\n",
      "1        2\n",
      "2        6\n",
      "3        5\n",
      "4        9\n",
      "        ..\n",
      "51322    1\n",
      "51323    3\n",
      "51324    1\n",
      "51325    1\n",
      "51326    1\n",
      "Name: transcript_count, Length: 51327, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# for numerical columns there may be a need to retrieve basic statistics\n",
    "transcript_column=df.transcript_count # df['transcript_count']\n",
    "print(transcript_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The maximum transcript_count observed:  170\n"
     ]
    }
   ],
   "source": [
    "# let's retrieve the statistics for the transcript_count column\n",
    "print(\"The maximum transcript_count observed: \", transcript_column.max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The minimum transcript_count observed:  1\n"
     ]
    }
   ],
   "source": [
    "print(\"The minimum transcript_count observed: \", transcript_column.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The average transcript_count observed:  3.700469538449549\n"
     ]
    }
   ],
   "source": [
    "print(\"The average transcript_count observed: \", transcript_column.mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The median of all transcript counts:  1.0\n"
     ]
    }
   ],
   "source": [
    "# the median is the middle element of a sorted sequence of elements\n",
    "# e.g. 1,1,3,4,103,2014,98531 -> median: 4\n",
    "print(\"The median of all transcript counts: \", transcript_column.median())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- by using the describe() method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    51327.000000\n",
       "mean         3.700470\n",
       "std          5.336323\n",
       "min          1.000000\n",
       "25%          1.000000\n",
       "50%          1.000000\n",
       "75%          4.000000\n",
       "max        170.000000\n",
       "Name: transcript_count, dtype: float64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pandas also provides the describe() method for statistics in numerical columns\n",
    "transcript_column.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plotting a histogram on a numerical column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='Frequency'>"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZUAAAD4CAYAAAAkRnsLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAY5ElEQVR4nO3dfbBcdZ3n8ffHhEF8AHkIbirBCUrWHbA0SmSpUmdVnCEyq8CMzsaaGlK77MRhsUprZqsEnRqZP1Ilu6vsMLvgwEIRWBUiPpB1ZFdER2uqkHhxeAoPy1VQYrIQhRGcUTT43T/6d7Vzcx868fTt2/B+VZ3qX3/7/E5/z6lOvvd3fqdPp6qQJKkLzxl1ApKkZw6LiiSpMxYVSVJnLCqSpM5YVCRJnVk66gQW2lFHHVWrVq0adRqSNFZuu+2271fVsvnWe9YVlVWrVjExMTHqNCRprCT5ziDrefpLktSZoRWVJM9Nsi3JHUm2J/mLFr8gyfeS3N6W0/r6nJ9kMsn9SU7ti5+Y5K722sVJ0uIHJ7muxW9NsmpY+yNJmt8wRypPAW+uqlcBa4B1SU5ur11UVWva8gWAJMcD64ETgHXAJUmWtPUvBTYCq9uyrsXPBh6vquOAi4ALh7g/kqR5DK2oVM+P2tOD2jLXPWFOB66tqqeq6kFgEjgpyXLg0Kq6pXr3lLkaOKOvz+bWvh44ZWoUI0laeEOdU0myJMntwKPATVV1a3vpPUnuTHJlksNbbAXwcF/3HS22orWnx/fqU1V7gB8CR86Qx8YkE0kmdu/e3c3OSZL2MdSiUlVPV9UaYCW9Uccr6J3Kehm9U2K7gI+01WcaYdQc8bn6TM/jsqpaW1Vrly2b94o4SdIBWpCrv6rqH4C/BdZV1SOt2PwcuBw4qa22Azimr9tKYGeLr5whvlefJEuBw4DHhrMXkqT5DPPqr2VJXtTahwBvAe5rcyRTzgTubu2twPp2Rdex9Cbkt1XVLuDJJCe3+ZKzgBv6+mxo7XcAXy7v5S9JIzPMLz8uBza3K7ieA2ypqs8nuSbJGnqnqR4C3g1QVduTbAHuAfYA51bV021b5wBXAYcAN7YF4ArgmiST9EYo64e4P5KkeeTZ9of92rVr60C/Ub/qvL/5RfuhD/9OVylJ0qKX5LaqWjvfen6jXpLUGYuKJKkzFhVJUmcsKpKkzlhUJEmdsahIkjpjUZEkdcaiIknqjEVFktQZi4okqTMWFUlSZywqkqTOWFQkSZ2xqEiSOmNRkSR1xqIiSeqMRUWS1BmLiiSpMxYVSVJnLCqSpM5YVCRJnRlaUUny3CTbktyRZHuSv2jxI5LclOSB9nh4X5/zk0wmuT/JqX3xE5Pc1V67OEla/OAk17X4rUlWDWt/JEnzG+ZI5SngzVX1KmANsC7JycB5wM1VtRq4uT0nyfHAeuAEYB1wSZIlbVuXAhuB1W1Z1+JnA49X1XHARcCFQ9wfSdI8hlZUqudH7elBbSngdGBzi28Gzmjt04Frq+qpqnoQmAROSrIcOLSqbqmqAq6e1mdqW9cDp0yNYiRJC2+ocypJliS5HXgUuKmqbgVeXFW7ANrj0W31FcDDfd13tNiK1p4e36tPVe0BfggcOUMeG5NMJJnYvXt3R3snSZpuqEWlqp6uqjXASnqjjlfMsfpMI4yaIz5Xn+l5XFZVa6tq7bJly+bJWpJ0oBbk6q+q+gfgb+nNhTzSTmnRHh9tq+0AjunrthLY2eIrZ4jv1SfJUuAw4LFh7IMkaX7DvPprWZIXtfYhwFuA+4CtwIa22gbghtbeCqxvV3QdS29Cfls7RfZkkpPbfMlZ0/pMbesdwJfbvIskaQSWDnHby4HN7Qqu5wBbqurzSW4BtiQ5G/gu8E6AqtqeZAtwD7AHOLeqnm7bOge4CjgEuLEtAFcA1ySZpDdCWT/E/ZEkzWNoRaWq7gRePUP8B8Aps/TZBGyaIT4B7DMfU1U/oRUlSdLo+Y16SVJnLCqSpM5YVCRJnbGoSJI6Y1GRJHXGoiJJ6oxFRZLUGYuKJKkzFhVJUmcsKpKkzlhUJEmdsahIkjpjUZEkdcaiIknqjEVFktQZi4okqTMWFUlSZywqkqTOWFQkSZ2xqEiSOmNRkSR1ZmhFJckxSb6S5N4k25O8t8UvSPK9JLe35bS+PucnmUxyf5JT++InJrmrvXZxkrT4wUmua/Fbk6wa1v5IkuY3zJHKHuBPq+o3gJOBc5Mc3167qKrWtOULAO219cAJwDrgkiRL2vqXAhuB1W1Z1+JnA49X1XHARcCFQ9wfSdI8hlZUqmpXVX2ztZ8E7gVWzNHldODaqnqqqh4EJoGTkiwHDq2qW6qqgKuBM/r6bG7t64FTpkYxkqSFtyBzKu201KuBW1voPUnuTHJlksNbbAXwcF+3HS22orWnx/fqU1V7gB8CR87w/huTTCSZ2L17dzc7JUnax9CLSpIXAJ8G3ldVT9A7lfUyYA2wC/jI1KozdK854nP12TtQdVlVra2qtcuWLdu/HZAkDWyoRSXJQfQKyser6jMAVfVIVT1dVT8HLgdOaqvvAI7p674S2NniK2eI79UnyVLgMOCx4eyNJGk+w7z6K8AVwL1V9dG++PK+1c4E7m7trcD6dkXXsfQm5LdV1S7gySQnt22eBdzQ12dDa78D+HKbd5EkjcDSIW77dcAfAnclub3FPgC8K8kaeqepHgLeDVBV25NsAe6hd+XYuVX1dOt3DnAVcAhwY1ugV7SuSTJJb4Syfoj7I0max9CKSlX9HTPPeXxhjj6bgE0zxCeAV8wQ/wnwzl8hTUlSh/xGvSSpMxYVSVJnLCqSpM5YVCRJnbGoSJI6Y1GRJHXGoiJJ6oxFRZLUGYuKJKkzFhVJUmcsKpKkzlhUJEmdsahIkjpjUZEkdWagopJkn9vOS5I03aAjlY8l2ZbkPyR50TATkiSNr4GKSlW9HvgDer8HP5HkE0l+a6iZSZLGzsBzKlX1APBnwPuBfwVcnOS+JL87rOQkSeNl0DmVVya5CLgXeDPwtqr6jda+aIj5SZLGyKC/Uf/fgMuBD1TVj6eCVbUzyZ8NJTNJ0tgZtKicBvy4qp4GSPIc4LlV9U9Vdc3QspMkjZVB51S+BBzS9/x5LTarJMck+UqSe5NsT/LeFj8iyU1JHmiPh/f1OT/JZJL7k5zaFz8xyV3ttYuTpMUPTnJdi9+aZNWA+yNJGoJBi8pzq+pHU09a+3nz9NkD/GmbezkZODfJ8cB5wM1VtRq4uT2nvbYeOAFYB1ySZEnb1qXARmB1W9a1+NnA41V1HL25nQsH3B9J0hAMWlT+Mclrpp4kORH48RzrU1W7quqbrf0kvUn+FcDpwOa22mbgjNY+Hbi2qp6qqgeBSeCkJMuBQ6vqlqoq4Oppfaa2dT1wytQoRpK08AadU3kf8KkkO9vz5cC/GfRN2mmpVwO3Ai+uql3QKzxJjm6rrQC+3tdtR4v9rLWnx6f6PNy2tSfJD4Ejge8PmpskqTsDFZWq+kaSfwG8HAhwX1X9bJC+SV4AfBp4X1U9McdAYqYXao74XH2m57CR3ukzXvKSl8yXsiTpAO3PDSVfC7yS3ojjXUnOmq9DkoPoFZSPV9VnWviRdkqL9vhoi++g9439KSuBnS2+cob4Xn2SLAUOAx6bnkdVXVZVa6tq7bJlywbYVUnSgRj0y4/XAP8FeD294vJaYO08fQJcAdxbVR/te2krsKG1NwA39MXXtyu6jqU3Ib+tnSp7MsnJbZtnTeszta13AF9u8y6SpBEYdE5lLXD8fv6H/TrgD4G7ktzeYh8APgxsSXI28F3gnQBVtT3JFuAeeleOnTv1vRjgHOAqepc139gW6BWta5JM0huhrN+P/CRJHRu0qNwN/DNg16Abrqq/Y+Y5D4BTZumzCdg0Q3wC2Of2+1X1E1pRkiSN3qBF5SjgniTbgKemglX19qFkJUkaS4MWlQuGmYQk6Zlh0EuKv5rk14HVVfWlJM8DlszXT5L07DLo1V9/RO8b63/dQiuAzw0pJ0nSmBr0eyrn0rua6wn4xQ92HT1nD0nSs86gReWpqvrp1JP2RUO/DyJJ2sugReWrST4AHNJ+m/5TwP8aXlqSpHE0aFE5D9gN3AW8G/gCvd+rlyTpFwa9+uvn9H5O+PLhpiNJGmcDFZUkDzLDHEpVvbTzjCRJY2t/7v015bn0bo1yRPfpSJLG2UBzKlX1g77le1X1X4E3Dzc1SdK4GfT012v6nj6H3sjlhUPJSJI0tgY9/fWRvvYe4CHg9zvPRpI01ga9+utNw05EkjT+Bj399SdzvT7tlx0lSc9S+3P112vp/XwvwNuArwEPDyMpSdJ42p8f6XpNVT0JkOQC4FNV9e+HlZgkafwMepuWlwA/7Xv+U2BV59lIksbaoCOVa4BtST5L75v1ZwJXDy0rSdJYGvTqr01JbgTe0EL/tqr+fnhpSZLG0aCnvwCeBzxRVX8J7Ehy7JBykiSNqUF/TvhDwPuB81voIOB/ztPnyiSPJrm7L3ZBku8lub0tp/W9dn6SyST3Jzm1L35ikrvaaxcnSYsfnOS6Fr81yaqB91qSNBSDjlTOBN4O/CNAVe1k/tu0XAWsmyF+UVWtacsXAJIcD6wHTmh9LkmypK1/KbARWN2WqW2eDTxeVccBFwEXDrgvkqQhGbSo/LSqinb7+yTPn69DVX0NeGzA7Z8OXFtVT1XVg8AkcFKS5cChVXVLe/+rgTP6+mxu7euBU6ZGMZKk0Ri0qGxJ8tfAi5L8EfAlDvwHu96T5M52euzwFlvB3l+k3NFiK1p7enyvPlW1B/ghcORMb5hkY5KJJBO7d+8+wLQlSfOZt6i0v/6vozca+DTwcuDPq+qvDuD9LgVeBqwBdvHLG1XONMKoOeJz9dk3WHVZVa2tqrXLli3br4QlSYOb95Liqqokn6uqE4GbfpU3q6pHptpJLgc+357uAI7pW3UlsLPFV84Q7++zI8lS4DAGP90mSRqCQU9/fT3Ja3/VN2tzJFPOBKauDNsKrG9XdB1Lb0J+W1XtAp5McnIbMZ0F3NDXZ0NrvwP4cpt3kSSNyKDfqH8T8MdJHqJ3BVjoDWJeOVuHJJ8E3ggclWQH8CHgjUnW0DtN9RDwbnob2p5kC3APvd9rObeqnm6bOofelWSHADe2BeAK4Jokk/RGKOsH3BdJ0pDMWVSSvKSqvgu8dX83XFXvmiF8xRzrbwI2zRCfAF4xQ/wnwDv3Ny9J0vDMN1L5HL27E38nyaer6vcWICdJ0piab06l/wqrlw4zEUnS+JuvqNQsbUmS9jHf6a9XJXmC3ojlkNaGX07UHzrU7CRJY2XOolJVS+Z6XZKkfvtz63tJkuZkUZEkdcaiIknqjEVFktQZi4okqTMWFUlSZywqkqTOWFQkSZ2xqEiSOmNRkSR1xqIiSeqMRUWS1BmLiiSpMxYVSVJnLCqSpM5YVCRJnRlaUUlyZZJHk9zdFzsiyU1JHmiPh/e9dn6SyST3Jzm1L35ikrvaaxcnSYsfnOS6Fr81yaph7YskaTDDHKlcBaybFjsPuLmqVgM3t+ckOR5YD5zQ+lySZOpXJy8FNgKr2zK1zbOBx6vqOOAi4MKh7YkkaSBDKypV9TXgsWnh04HNrb0ZOKMvfm1VPVVVDwKTwElJlgOHVtUtVVXA1dP6TG3reuCUqVGMJGk0FnpO5cVVtQugPR7d4iuAh/vW29FiK1p7enyvPlW1B/ghcORMb5pkY5KJJBO7d+/uaFckSdMtlon6mUYYNUd8rj77Bqsuq6q1VbV22bJlB5iiJGk+C11UHmmntGiPj7b4DuCYvvVWAjtbfOUM8b36JFkKHMa+p9skSQtooYvKVmBDa28AbuiLr29XdB1Lb0J+WztF9mSSk9t8yVnT+kxt6x3Al9u8iyRpRJYOa8NJPgm8ETgqyQ7gQ8CHgS1Jzga+C7wToKq2J9kC3APsAc6tqqfbps6hdyXZIcCNbQG4ArgmySS9Ecr6Ye2LJGkwQysqVfWuWV46ZZb1NwGbZohPAK+YIf4TWlGSJC0Oi2WiXpL0DGBRkSR1xqIiSeqMRUWS1BmLiiSpMxYVSVJnLCqSpM5YVCRJnbGoSJI6Y1GRJHXGoiJJ6oxFRZLUGYuKJKkzFhVJUmcsKpKkzlhUJEmdsahIkjpjUZEkdcaiIknqjEVFktQZi4okqTMjKSpJHkpyV5Lbk0y02BFJbkryQHs8vG/985NMJrk/yal98RPbdiaTXJwko9gfSVLPKEcqb6qqNVW1tj0/D7i5qlYDN7fnJDkeWA+cAKwDLkmypPW5FNgIrG7LugXMX5I0zWI6/XU6sLm1NwNn9MWvraqnqupBYBI4Kcly4NCquqWqCri6r48kaQRGVVQK+GKS25JsbLEXV9UugPZ4dIuvAB7u67ujxVa09vT4PpJsTDKRZGL37t0d7oYkqd/SEb3v66pqZ5KjgZuS3DfHujPNk9Qc8X2DVZcBlwGsXbt2xnUkSb+6kRSVqtrZHh9N8lngJOCRJMuralc7tfVoW30HcExf95XAzhZfOUN8Qaw6729+0X7ow7+zUG8rSYvagp/+SvL8JC+cagO/DdwNbAU2tNU2ADe09lZgfZKDkxxLb0J+WztF9mSSk9tVX2f19ZEkjcAoRiovBj7brv5dCnyiqv53km8AW5KcDXwXeCdAVW1PsgW4B9gDnFtVT7dtnQNcBRwC3NgWSdKILHhRqapvA6+aIf4D4JRZ+mwCNs0QnwBe0XWOkqQDs5guKZYkjTmLiiSpMxYVSVJnLCqSpM5YVCRJnRnVN+qfUfwipCT1OFKRJHXGoiJJ6oxFRZLUGYuKJKkzFhVJUmcsKpKkznhJ8RD1X2oMXm4s6ZnPkYokqTMWFUlSZzz9tYD85r2kZzpHKpKkzjhS6dj0yXlJejZxpCJJ6owjlUXAuRZJzxSOVCRJnXGkMiKzzb04apE0zsa+qCRZB/wlsAT4H1X14RGnNBQWG0njYKyLSpIlwH8HfgvYAXwjydaqume0mXXD0YykcTPWRQU4CZisqm8DJLkWOB14RhSVQezvJcz9RcjiJKlr415UVgAP9z3fAfzL6Ssl2QhsbE9/lOT+/Xyfo4DvH1CGo7VP3rlw5hVni4/AOB5rc14Y5rwwZsv51wfpPO5FJTPEap9A1WXAZQf8JslEVa090P6jMo55m/PCMOeF8WzMedwvKd4BHNP3fCWwc0S5SNKz3rgXlW8Aq5Mcm+TXgPXA1hHnJEnPWmN9+quq9iR5D/B/6F1SfGVVbR/CWx3wqbMRG8e8zXlhmPPCeNblnKp9piAkSTog4376S5K0iFhUJEmdsajMI8m6JPcnmUxy3qjzmUmSY5J8Jcm9SbYneW+LX5Dke0lub8tpo861X5KHktzVcptosSOS3JTkgfZ4+KjznJLk5X3H8vYkTyR532I7zkmuTPJokrv7YrMe1yTnt8/3/UlOXUQ5/+ck9yW5M8lnk7yoxVcl+XHf8f7YIsp51s/CIj7O1/Xl+1CS21v8wI5zVbnMstCb/P8W8FLg14A7gONHndcMeS4HXtPaLwT+L3A8cAHwH0ed3xx5PwQcNS32n4DzWvs84MJR5znHZ+P/0ftC2KI6zsBvAq8B7p7vuLbPyR3AwcCx7fO+ZJHk/NvA0ta+sC/nVf3rLbLjPONnYTEf52mvfwT481/lODtSmdsvbgNTVT8Fpm4Ds6hU1a6q+mZrPwncS+9uA+PodGBza28GzhhdKnM6BfhWVX1n1IlMV1VfAx6bFp7tuJ4OXFtVT1XVg8Akvc/9gpop56r6YlXtaU+/Tu97aIvGLMd5Nov2OE9JEuD3gU/+Ku9hUZnbTLeBWdT/WSdZBbwauLWF3tNOH1y5mE4lNQV8Mclt7VY6AC+uql3QK5bA0SPLbm7r2fsf32I+zjD7cR2Xz/i/A27se35skr9P8tUkbxhVUrOY6bMwDsf5DcAjVfVAX2y/j7NFZW4D3QZmsUjyAuDTwPuq6gngUuBlwBpgF72h7WLyuqp6DfBW4NwkvznqhAbRvmj7duBTLbTYj/NcFv1nPMkHgT3Ax1toF/CSqno18CfAJ5IcOqr8ppnts7DojzPwLvb+Q+mAjrNFZW5jcxuYJAfRKygfr6rPAFTVI1X1dFX9HLicEQy351JVO9vjo8Bn6eX3SJLlAO3x0dFlOKu3At+sqkdg8R/nZrbjuqg/40k2AP8a+INqJ/rbKaQftPZt9OYn/vnosvylOT4Li/04LwV+F7huKnagx9miMrexuA1MOxd6BXBvVX20L768b7Uzgbun9x2VJM9P8sKpNr1J2bvpHd8NbbUNwA2jyXBOe/1Ft5iPc5/ZjutWYH2Sg5McC6wGto0gv32k9wN87wfeXlX/1Bdflt5vKZHkpfRy/vZostzbHJ+FRXucm7cA91XVjqnAAR/nhb76YNwW4DR6V1N9C/jgqPOZJcfX0xtK3wnc3pbTgGuAu1p8K7B81Ln25fxSelfD3AFsnzq2wJHAzcAD7fGIUec6Le/nAT8ADuuLLarjTK/g7QJ+Ru8v5LPnOq7AB9vn+37grYso50l68xBTn+mPtXV/r31m7gC+CbxtEeU862dhsR7nFr8K+ONp6x7QcfY2LZKkznj6S5LUGYuKJKkzFhVJUmcsKpKkzlhUJEmdsahIkjpjUZEkdeb/A0qQowkthLf3AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "transcript_column.plot(kind=\"hist\",bins=100)\n",
    "# bins: refers to the number of bars that appear (configures the histogram resolution)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sorting DataFrame rows\n",
    "\n",
    "Sorted by values of one (or more) columns\n",
    "\n",
    "Sorting returns a new sorted DataFrame\n",
    "\n",
    "## Sorting order\n",
    "\n",
    "- ascending"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47550</th>\n",
       "      <td>PCBP1-AS1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32044</th>\n",
       "      <td>LINC00969</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30080</th>\n",
       "      <td>TEX41</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16363</th>\n",
       "      <td>TCF4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr18</td>\n",
       "      <td>-</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11524</th>\n",
       "      <td>NDRG2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr14</td>\n",
       "      <td>-</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27712</th>\n",
       "      <td>RP11-986E7.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chr14</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27711</th>\n",
       "      <td>RP11-305B6.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chr14</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27710</th>\n",
       "      <td>RP11-108M12.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chr14</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27709</th>\n",
       "      <td>RP11-545M17.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51326</th>\n",
       "      <td>BCYRN1</td>\n",
       "      <td>scRNA</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51327 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name          gene_biotype chromosome strand  transcript_count\n",
       "47550      PCBP1-AS1  processed_transcript       chr2      -               170\n",
       "32044      LINC00969               lincRNA       chr3      +               142\n",
       "30080          TEX41               lincRNA       chr2      +               103\n",
       "16363           TCF4        protein_coding      chr18      -                90\n",
       "11524          NDRG2        protein_coding      chr14      -                79\n",
       "...              ...                   ...        ...    ...               ...\n",
       "27712   RP11-986E7.2  processed_pseudogene      chr14      -                 1\n",
       "27711   RP11-305B6.1  processed_pseudogene      chr14      -                 1\n",
       "27710  RP11-108M12.2  processed_pseudogene      chr14      -                 1\n",
       "27709  RP11-545M17.2  processed_pseudogene      chr14      +                 1\n",
       "51326         BCYRN1                 scRNA       chr2      +                 1\n",
       "\n",
       "[51327 rows x 5 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sorting DataFrame, by alphabetical order of gene_name\n",
    "df.sort_values('transcript_count', ascending=False)\n",
    "#df.sort_values('transcript_count', ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- descending"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47550</th>\n",
       "      <td>PCBP1-AS1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32044</th>\n",
       "      <td>LINC00969</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30080</th>\n",
       "      <td>TEX41</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16363</th>\n",
       "      <td>TCF4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr18</td>\n",
       "      <td>-</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11524</th>\n",
       "      <td>NDRG2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr14</td>\n",
       "      <td>-</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17421</th>\n",
       "      <td>ADGRG1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14712</th>\n",
       "      <td>CACNB4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47657</th>\n",
       "      <td>SATB1-AS1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37157</th>\n",
       "      <td>AC012594.1</td>\n",
       "      <td>antisense</td>\n",
       "      <td>chr2</td>\n",
       "      <td>-</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37845</th>\n",
       "      <td>TTN-AS1</td>\n",
       "      <td>antisense</td>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name          gene_biotype chromosome strand  transcript_count\n",
       "47550   PCBP1-AS1  processed_transcript       chr2      -               170\n",
       "32044   LINC00969               lincRNA       chr3      +               142\n",
       "30080       TEX41               lincRNA       chr2      +               103\n",
       "16363        TCF4        protein_coding      chr18      -                90\n",
       "11524       NDRG2        protein_coding      chr14      -                79\n",
       "17421      ADGRG1        protein_coding      chr16      +                77\n",
       "14712      CACNB4        protein_coding       chr2      -                74\n",
       "47657   SATB1-AS1  processed_transcript       chr3      +                73\n",
       "37157  AC012594.1             antisense       chr2      -                73\n",
       "37845     TTN-AS1             antisense       chr2      +                71"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sorting DataFrame, by the number of transcripts of each gene, in descending order\n",
    "df.sort_values('transcript_count', ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Grouping examples\n",
    "\n",
    "* For each biotype, find the average number of transcripts associated to genes belonging to the biotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr20</td>\n",
       "      <td>-</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FGR</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CFH</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FUCA2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GCLC</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NFYA</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr6</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name    gene_biotype chromosome strand  transcript_count\n",
       "0    TSPAN6  protein_coding       chrX      -                 5\n",
       "1      TNMD  protein_coding       chrX      +                 2\n",
       "2      DPM1  protein_coding      chr20      -                 6\n",
       "3     SCYL3  protein_coding       chr1      -                 5\n",
       "4  C1orf112  protein_coding       chr1      +                 9\n",
       "5       FGR  protein_coding       chr1      -                 7\n",
       "6       CFH  protein_coding       chr1      +                 6\n",
       "7     FUCA2  protein_coding       chr6      -                 3\n",
       "8      GCLC  protein_coding       chr6      -                13\n",
       "9      NFYA  protein_coding       chr6      +                 2"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# call a head() just to remember how df looks like... \n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_biotype\n",
      "3prime_overlapping_ncRNA              1.153846\n",
      "IG_C_gene                             1.285714\n",
      "IG_C_pseudogene                       1.000000\n",
      "IG_D_gene                             1.000000\n",
      "IG_J_gene                             1.000000\n",
      "IG_J_pseudogene                       1.000000\n",
      "IG_V_gene                             1.053846\n",
      "IG_V_pseudogene                       1.000000\n",
      "Mt_rRNA                               1.000000\n",
      "Mt_tRNA                               1.000000\n",
      "TEC                                   1.166667\n",
      "TR_C_gene                             1.800000\n",
      "TR_D_gene                             1.000000\n",
      "TR_J_gene                             1.000000\n",
      "TR_J_pseudogene                       1.000000\n",
      "TR_V_gene                             1.010753\n",
      "TR_V_pseudogene                       1.000000\n",
      "antisense                             1.871659\n",
      "bidirectional_promoter_lncRNA         2.000000\n",
      "lincRNA                               1.690283\n",
      "miRNA                                 1.000000\n",
      "misc_RNA                              1.000000\n",
      "polymorphic_pseudogene                3.206897\n",
      "processed_pseudogene                  1.000304\n",
      "processed_transcript                  5.735556\n",
      "protein_coding                        7.439154\n",
      "pseudogene                            1.000000\n",
      "rRNA                                  1.000000\n",
      "ribozyme                              1.000000\n",
      "scRNA                                 1.000000\n",
      "scaRNA                                1.000000\n",
      "sense_intronic                        1.087464\n",
      "sense_overlapping                     1.609195\n",
      "snRNA                                 1.000000\n",
      "snoRNA                                1.027228\n",
      "transcribed_processed_pseudogene      2.100939\n",
      "transcribed_unitary_pseudogene        5.056604\n",
      "transcribed_unprocessed_pseudogene    3.337368\n",
      "unitary_pseudogene                    1.921986\n",
      "unprocessed_pseudogene                1.001364\n",
      "vaultRNA                              1.000000\n",
      "Name: transcript_count, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# Constructing the command:\n",
    "#\n",
    "# Dataframe: \n",
    "#           df\n",
    "#\n",
    "# Groub by gene_biotype: \n",
    "#           df.groupby('gene_biotype')\n",
    "#\n",
    "\n",
    "result = df.groupby('gene_biotype')['transcript_count'].mean()\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='Frequency'>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAD4CAYAAADrRI2NAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAPG0lEQVR4nO3db4xmZX3G8e8FS4MgBsyOdIPoKiEoMQg4YJtVi1IasYrSxlbSWmKsa1NsIJpUJKbSF018oWCbNtZFqKiIFRGlxlqRqmhSxVmkAi4GtagLW3Zs0+xijRT89cWcbab7b87MznnOPnt/P8lkzrmfP+faYbjmzD33c55UFZKkdhw2dgBJ0mRZ/JLUGItfkhpj8UtSYyx+SWrMmrED9LF27dpav3792DEkaaps3rz5J1U1s/v4VBT/+vXrmZubGzuGJE2VJD/c27hTPZLUGItfkhpj8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1JipeOXugfj+zIv3On7S/FcnnESSDg6e8UtSYyx+SWqMxS9JjbH4JakxgxV/khOTfCnJliT3Jbm0G78yyUNJ7u4+XjFUBknSnoZc1fM48LaquivJMcDmJLd1t11dVe8Z8NiSpH0YrPirahuwrdvemWQLcMJQx5Mk9TOROf4k64EzgG90Q29J8u0k1yU5bh+P2ZhkLsnc/Pz8JGJKUhMGL/4kTwZuBi6rqh3A+4GTgNNZ+I3gvXt7XFVtqqrZqpqdmdnjLSMlSSs0aPEnOYKF0r+hqj4FUFWPVNUTVfUL4Brg7CEzSJL+vyFX9QS4FthSVVctGl+36G4XAvcOlUGStKchV/VsAF4P3JPk7m7sCuCiJKcDBTwIvHnADJKk3Qy5qudrQPZy0+eGOqYkaWm+cleSGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1BiLX5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pekxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1ZrDiT3Jiki8l2ZLkviSXduNPTXJbkge6z8cNlUGStKchz/gfB95WVc8FfgW4JMmpwOXA7VV1MnB7ty9JmpDBir+qtlXVXd32TmALcALwauD67m7XA68ZKoMkaU8TmeNPsh44A/gGcHxVbYOFHw7A0yaRQZK0YPDiT/Jk4GbgsqrasYzHbUwyl2Rufn5+uICS1JhBiz/JESyU/g1V9alu+JEk67rb1wHb9/bYqtpUVbNVNTszMzNkTElqypCregJcC2ypqqsW3XQrcHG3fTHwmaEySJL2tGbA594AvB64J8nd3dgVwLuBTyR5I/Aj4LUDZpAk7Waw4q+qrwHZx83nDnVcSdL++cpdSWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1BiLX5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pekxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUmF7Fn+R5QweRJE1G3zP+v01yZ5I/TnLskIEkScPqVfxV9SLg94ATgbkkH0ty3qDJJEmD6D3HX1UPAO8E3g78GvBXSe5P8ltDhZMkrb6+c/ynJbka2AK8DHhVVT232756wHySpFW2puf9/hq4Briiqn62a7CqHk7yzkGSSZIG0bf4XwH8rKqeAEhyGHBkVf13VX1ksHSSpFXXd47/i8CTFu0f1Y1JkqZM3+I/sqoe3bXTbR81TCRJ0pD6Fv9Pk5y5ayfJC4Cf7ef+JLkuyfYk9y4auzLJQ0nu7j5esbLYkqSV6jvHfxlwU5KHu/11wO8u8ZgPsfBH4Q/vNn51Vb2nb0BJ0urqVfxV9c0kzwFOAQLcX1X/s8Rj7kiy/sAjSpJW03Iu0nYWcBpwBnBRkj9Y4THfkuTb3VTQcfu6U5KNSeaSzM3Pz6/wUJKk3fV9AddHgPcAL2LhB8BZwOwKjvd+4CTgdGAb8N593bGqNlXVbFXNzszMrOBQkqS96TvHPwucWlV1IAerqkd2bSe5BvjsgTyfJGn5+k713Av88oEeLMm6RbsXds8rSZqgvmf8a4HvJLkT+Pmuwaq6YF8PSHIjcA6wNslW4F3AOUlOBwp4EHjzilJLklasb/FfudwnrqqL9jJ87XKfR5K0uvou5/xKkmcCJ1fVF5McBRw+bDRJ0hD6rup5E/BJ4APd0AnApwfKJEkaUN8/7l4CbAB2wP+9KcvThgolSRpO3+L/eVU9tmsnyRoW/kArSZoyfYv/K0muAJ7UvdfuTcA/DBdLkjSUvsV/OTAP3MPCEszPsfD+u5KkKdN3Vc8vWHjrxWuGjSNJGlqv4k/yb+xlTr+qnr3qiSRJg1rOtXp2ORJ4LfDU1Y8jSRparzn+qvqPRR8PVdX7gJcNG02SNIS+Uz1nLto9jIXfAI4ZJJEkaVB9p3oWXzf/cRYusPY7q55GkjS4vqt6Xjp0EEnSZPSd6nnr/m6vqqtWJ44kaWjLWdVzFnBrt/8q4A7gx0OEkiQNZzlvxHJmVe0ESHIlcFNV/eFQwSRJw+h7yYZnAI8t2n8MWL/qaSRJg+t7xv8R4M4kt7DwCt4LgQ8PlkqSNJi+q3r+Isk/Ai/uht5QVd8aLpYkaSh9p3oAjgJ2VNVfAluTPGugTJKkAfV968V3AW8H3tENHQF8dKhQkqTh9D3jvxC4APgpQFU9jJdskKSp1Lf4H6uqors0c5Kjh4skSRpS3+L/RJIPAMcmeRPwRXxTFkmaSkuu6kkS4O+B5wA7gFOAP6uq2wbOJkkawJLFX1WV5NNV9QLAspekKdd3qufrSc4aNIkkaSL6vnL3pcAfJXmQhZU9YeGXgdOGCiZJGsZ+iz/JM6rqR8D5E8ojSRrYUmf8n2bhqpw/THJzVf32BDJJkga01Bx/Fm0/e8ggkqTJWKr4ax/bS0pyXZLtSe5dNPbUJLcleaD7fNxynlOSdOCWKv7nJ9mRZCdwWre9I8nOJDuWeOyHgJfvNnY5cHtVnQzc3u1LkiZov3P8VXX4Sp+4qu5Isn634VcD53Tb1wNfZuHib5KkCVnOZZlXw/FVtQ2g+/y0fd0xycYkc0nm5ufnJxZQkg51ky7+3qpqU1XNVtXszMzM2HEk6ZAx6eJ/JMk6gO7z9gkfX5KaN+nivxW4uNu+GPjMhI8vSc0brPiT3Aj8C3BKkq1J3gi8GzgvyQPAed2+JGmC+l6rZ9mq6qJ93HTuUMeUJC3toP3jriRpGBa/JDXG4pekxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1BiLX5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pekxlj8ktQYi1+SGmPxS1JjLH5JasyaMQ6a5EFgJ/AE8HhVzY6RQ5JaNErxd15aVT8Z8fiS1CSneiSpMWMVfwFfSLI5yca93SHJxiRzSebm5+cnHE+SDl1jFf+GqjoTOB+4JMlLdr9DVW2qqtmqmp2ZmZl8Qkk6RI1S/FX1cPd5O3ALcPYYOSSpRRMv/iRHJzlm1zbwG8C9k84hSa0aY1XP8cAtSXYd/2NV9fkRckhSkyZe/FX1A+D5kz6uJGmByzklqTEWvyQ1xuKXpMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1BiLX5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDVmjPfcPSh8f+bFex0/af6rox17XyaRSVI7POOXpMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4Jakxza7j35eVrO9f7rr85RrzNQdD8zUNq2+s75dD+ft0aPv7/2CIr59n/JLUGItfkhpj8UtSYyx+SWrMKMWf5OVJvpvke0kuHyODJLVq4sWf5HDgb4DzgVOBi5KcOukcktSqMc74zwa+V1U/qKrHgI8Drx4hhyQ1aYx1/CcAP160vxV44e53SrIR2NjtPprkuys83lrgJyt87OJAB/wUK7D/7ONkWo7V+dovNrl/8+pnn5y9Zx/r+2X5xz30vvYH4sD+uz1zb4NjFP/e/hW1x0DVJmDTAR8smauq2QN9njFMc3aY7vxmH88055+W7GNM9WwFTly0/3Tg4RFySFKTxij+bwInJ3lWkl8CXgfcOkIOSWrSxKd6qurxJG8B/gk4HLiuqu4b8JAHPF00omnODtOd3+zjmeb8U5E9VXtMr0uSDmG+cleSGmPxS1JjDtniT3Jdku1J7h07y3IlOTHJl5JsSXJfkkvHztRXkiOT3JnkX7vsfz52puVKcniSbyX57NhZlivJg0nuSXJ3krmx8yxHkmOTfDLJ/d33/q+OnamPJKd0X+9dHzuSXDZ2rv05ZOf4k7wEeBT4cFU9b+w8y5FkHbCuqu5KcgywGXhNVX1n5GhLShLg6Kp6NMkRwNeAS6vq6yNH6y3JW4FZ4ClV9cqx8yxHkgeB2aqauhdAJbke+GpVfbBb8XdUVf3XyLGWpbskzUPAC6vqh2Pn2ZdD9oy/qu4A/nPsHCtRVduq6q5ueyewhYVXPB/0asGj3e4R3cfUnF0keTrwm8AHx87SkiRPAV4CXAtQVY9NW+l3zgW+fzCXPhzCxX+oSLIeOAP4xshReuumSu4GtgO3VdXUZAfeB/wp8IuRc6xUAV9Isrm77Mm0eDYwD/xdN832wSRHjx1qBV4H3Dh2iKVY/AexJE8GbgYuq6odY+fpq6qeqKrTWXhV9tlJpmKqLckrge1VtXnsLAdgQ1WdycLVby/ppjynwRrgTOD9VXUG8FNgqi7Z3k1PXQDcNHaWpVj8B6lufvxm4Iaq+tTYeVai+1X9y8DLx03S2wbggm6e/OPAy5J8dNxIy1NVD3eftwO3sHA13GmwFdi66LfDT7Lwg2CanA/cVVWPjB1kKRb/Qaj7A+m1wJaqumrsPMuRZCbJsd32k4BfB+4fNVRPVfWOqnp6Va1n4Vf2f66q3x85Vm9Jju4WA9BNk/wGMBWr2qrq34EfJzmlGzoXOOgXM+zmIqZgmgfGuTrnRCS5ETgHWJtkK/Cuqrp23FS9bQBeD9zTzZUDXFFVnxsvUm/rgOu71Q2HAZ+oqqlbFjmljgduWThvYA3wsar6/LiRluVPgBu6KZMfAG8YOU9vSY4CzgPePHaWPg7Z5ZySpL1zqkeSGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMb8L4V2sKrd0KF9AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "result.plot(kind=\"hist\",bins=50,color=\"Crimson\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
